pkg/volume/csi/csi_block.go

```diff
@@ -279,46 +200,121 @@ func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, vol
+		return "", errors.New("no existing VolumeAttachment found")
+	}
+
+	// Call NodeStageVolume
+	stagingPath, err := m.stageVolumeForBlock(csiSource, attachment)
+	if err != nil {
+		return "", err
+	}
+
+	// Call NodePublishVolume
+	publishPath, err := m.publishVolumeForBlock(csiSource, attachment, stagingPath)
+	if err != nil {
+		return "", err
+	}
+
+	return publishPath, nil
+}
+
+func (m *csiBlockMapper) MapDevice(devicePath, globalMapPath, volumeMapPath, volumeMapName string, podUID types.UID) error {
+	return ioutil.MapBlockVolume(devicePath, globalMapPath, volumeMapPath, volumeMapName, podUID)
+}
+
+var _ volume.BlockVolumeUnmapper = &csiBlockMapper{}
+
+func (m *csiBlockMapper) unpublishVolumeForBlock(publishPath string) error {
 	csi := m.csiClient
 	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
 	defer cancel()
 
-	// unmap global device map path
-	if err := csi.NodeUnstageVolume(ctx, m.volumeID, globalMapPath); err != nil {
-		glog.Errorf(log("blockMapper.TearDownDevice failed: %v", err))
+	// Request to remove symlink publishPath.
+	// If driver doesn't implement NodeUnstageVolume, detaching the device from the node is required, here.
+	if err := csi.NodeUnpublishVolume(ctx, m.volumeID, publishPath); err != nil {
+		glog.Error(log("blockMapper.unpublishVolumeForBlock failed: %v", err))
 		return err
 	}
-	glog.V(4).Infof(log("blockMapper.TearDownDevice NodeUnstageVolume successfully [%s]", globalMapPath))
+	glog.V(4).Infof(log("blockMapper.unpublishVolumeForBlock NodeUnpublished successfully [%s]", publishPath))
+
+	return nil
+}
+
+func (m *csiBlockMapper) unstageVolumeForBlock(stagingPath string) error {
+	csi := m.csiClient
+	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
+	defer cancel()
 
-	// request to remove pod volume map path also
-	podVolumePath, volumeName := m.GetPodDeviceMapPath()
-	podVolumeMapPath := filepath.Join(podVolumePath, volumeName)
-	if err := csi.NodeUnpublishVolume(ctx, m.volumeID, podVolumeMapPath); err != nil {
-		glog.Error(log("blockMapper.TearDownDevice failed: %v", err))
+	// Request to remove symlink stagingPath and to detach the device from the node.
+	if err := csi.NodeUnstageVolume(ctx, m.volumeID, stagingPath); err != nil {
+		glog.Errorf(log("blockMapper.unstageVolumeForBlock failed: %v", err))
 		return err
 	}
+	glog.V(4).Infof(log("blockMapper.unstageVolumeForBlock NodeUnstageVolume successfully [%s]", stagingPath))
+
+	return nil
+}
+
+// TearDownDevice removes traces of the SetUpDevice.
+func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error {
+	if !m.plugin.blockEnabled {
+		return errors.New("CSIBlockVolume feature not enabled")
+	}
 
-	glog.V(4).Infof(log("blockMapper.TearDownDevice NodeUnpublished successfully [%s]", podVolumeMapPath))
+	glog.V(4).Infof(log("unmapper.TearDownDevice(globalMapPath=%s; devicePath=%s)", globalMapPath, devicePath))
+
```

Comment:
Same here. Since calls for `unpublishVolumeForBlock` and `unstageVolumeForBlock` happen successively, create a `csiClient` here and pass it to each method as a parameter respectively to avoid creating client inside each call.